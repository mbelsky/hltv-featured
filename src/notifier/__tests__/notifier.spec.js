const { _notify } = require('../notifier')

const fetchDataFb = (users) => async () => {
  return {
    matches: [
      {
        stars: 2,
        id: 2341452,
        event: 'CLUTCH Season 2',
        href:
          'https://www.hltv.org/matches/2341452/w7m-vs-red-canids-clutch-season-2',
        title: 'W7M vs RED Canids',
        unixTimestamp: 1589749051000,
      },
    ],
    timeZoneOffsetsMap: {
      'Europe/Moscow': 10800,
    },
    users,
  }
}
const log = () => {}
const telegram = {}
const updateUser = jest.fn()
let processEnv

beforeAll(() => {
  processEnv = process.env
})

beforeEach(() => {
  process.env.NODE_ENV = 'production'
  telegram.sendMessage = jest.fn()
})

afterAll(() => {
  process.env = processEnv
})

describe('_notify', () => {
  test('send featured match message', async () => {
    const fetchData = fetchDataFb({
      0: {
        filter: 1,
        location: {
          timeZoneId: 'Europe/Moscow',
        },
      },
    })

    await _notify({
      fetchData,
      log,
      telegram,
      updateUser,
    })

    expect(telegram.sendMessage.mock.calls.length).toBe(1)
    expect(telegram.sendMessage.mock.calls[0]).toMatchSnapshot()
  })

  test('send empty featured match message', async () => {
    const fetchData = fetchDataFb({
      1: {
        filter: 4,
      },
    })

    await _notify({
      fetchData,
      log,
      telegram,
      updateUser,
    })

    expect(telegram.sendMessage.mock.calls.length).toBe(1)
    expect(telegram.sendMessage.mock.calls[0]).toMatchSnapshot()
  })

  test('send favorite team match message', async () => {
    const fetchData = fetchDataFb({
      2: {
        filter: 4,
        favorites: ['w7m'],
        seenCustomizeLocationMessage: true,
      },
    })

    await _notify({
      fetchData,
      log,
      telegram,
      updateUser,
    })

    expect(telegram.sendMessage.mock.calls.length).toBe(1)
    expect(telegram.sendMessage.mock.calls[0]).toMatchSnapshot()
  })

  test('does not create & send favorite team match message', async () => {
    const fetchData = fetchDataFb({
      2: {
        filter: 4,
        favorites: ['navi'],
        seenEmptyMessage: true,
      },
    })

    await _notify({
      fetchData,
      log,
      telegram,
      updateUser,
    })

    expect(telegram.sendMessage.mock.calls.length).toBe(0)
  })

  test('does not send empty message second time for same users', async () => {
    const fetchData = fetchDataFb({
      2: {
        filter: 4,
        seenEmptyMessage: true,
      },
    })

    await _notify({
      fetchData,
      log,
      telegram,
      updateUser,
    })

    expect(telegram.sendMessage.mock.calls.length).toBe(0)
  })

  test('send custom location message', async () => {
    const fetchData = fetchDataFb({
      3: {
        filter: 0,
      },
    })

    await _notify({
      fetchData,
      log,
      telegram,
      updateUser,
    })

    expect(telegram.sendMessage.mock.calls.length).toBe(2)
    expect(telegram.sendMessage.mock.calls[1]).toMatchSnapshot()
  })

  test('create and do not send custom location message', async () => {
    const fetchData = fetchDataFb({
      3: {
        filter: 5,
      },
    })

    await _notify({
      fetchData,
      log,
      telegram,
      updateUser,
    })

    expect(telegram.sendMessage.mock.calls.length).toBe(1)
    expect(telegram.sendMessage.mock.calls[0]).toMatchSnapshot()
  })

  describe('dev enviroment', () => {
    const fetchData = fetchDataFb({
      128526: {
        filter: 1,
        location: {
          timeZoneId: 'Europe/Moscow',
        },
      },
    })

    beforeEach(() => {
      process.env = {}
    })

    test('send a featured match message for dev enviroment', async () => {
      process.env.TELEGRAM_USER_TEST_IDS = '128526'

      await _notify({
        fetchData,
        log,
        telegram,
        updateUser,
      })

      expect(telegram.sendMessage.mock.calls.length).toBe(1)
    })

    test('does not send a featured match message for dev enviroment', async () => {
      await _notify({
        fetchData,
        log,
        telegram,
        updateUser,
      })

      expect(telegram.sendMessage.mock.calls.length).toBe(0)
    })
  })

  test('splits long telegram messages', async () => {
    const fetchData = async () => {
      const match = {
        stars: 2,
        id: 2341452,
        event: 'CLUTCH Season 2',
        href:
          'https://www.hltv.org/matches/2341452/w7m-vs-red-canids-clutch-season-2',
        title: 'W7M vs RED Canids',
        unixTimestamp: 1589749051000,
      }
      const matches = []

      while (matches.length < 100) {
        matches.push(match)
      }

      return {
        matches,
        timeZoneOffsetsMap: {
          'Europe/Moscow': 10800,
        },
        users: {
          0: {
            filter: 1,
            location: {
              timeZoneId: 'Europe/Moscow',
            },
          },
        },
      }
    }

    await _notify({
      fetchData,
      log,
      telegram,
      updateUser,
    })

    expect(telegram.sendMessage.mock.calls.length).toBe(2)
  })
})
