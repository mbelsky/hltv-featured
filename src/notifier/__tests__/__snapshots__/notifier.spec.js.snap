// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_notify create and do not send custom location message 1`] = `
Array [
  3,
  "There are no 5-stars matches today. And I won't spam you with this message every morning.

Next time just send /upcoming if no morning notification.",
  Object {
    "disable_notification": true,
    "disable_web_page_preview": true,
    "parse_mode": "HTML",
  },
]
`;

exports[`_notify send custom location message 1`] = `
Array [
  3,
  "You may change matches timezone for your feed. Just tap on the ðŸ“Ž attachment button, choose Location and send your city coordinates as a message.",
  undefined,
]
`;

exports[`_notify send empty featured match message 1`] = `
Array [
  1,
  "There are no 4-stars matches today. And I won't spam you with this message every morning.

Next time just send /upcoming if no morning notification.",
  Object {
    "disable_notification": true,
    "disable_web_page_preview": true,
    "parse_mode": "HTML",
  },
]
`;

exports[`_notify send favorite team match message 1`] = `
Array [
  2,
  "<b>Your favorite teams matches:</b>

<a href=\\"https://www.hltv.org/matches/2341452/w7m-vs-red-canids-clutch-season-2\\">W7MÂ vsÂ REDÂ Canids</a>
Rating: â˜†â˜†
<i>17Â MayÂ 20:57Â UTC @ CLUTCH Season 2</i>",
  Object {
    "disable_web_page_preview": true,
    "parse_mode": "HTML",
  },
]
`;

exports[`_notify send featured match message 1`] = `
Array [
  0,
  "<a href=\\"https://www.hltv.org/matches/2341452/w7m-vs-red-canids-clutch-season-2\\">W7MÂ vsÂ REDÂ Canids</a>
Rating: â˜†â˜†
<i>17Â MayÂ 23:57 @ CLUTCH Season 2</i>",
  Object {
    "disable_web_page_preview": true,
    "parse_mode": "HTML",
  },
]
`;
